pkg/util/fileutil/writer.go

```diff
@@ -0,0 +1,119 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Forked from kubernetes/pkg/volume/util/atomic_writer.go @fcee7a0105
+// to borrow the file/path validation. This does not write files atomically
+// as described in https://github.com/kubernetes/kubernetes/issues/18372 so
+// behavior may be different from watching or reloading
+// files when compared to K8S native Secrets.
+
+package fileutil
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"sigs.k8s.io/secrets-store-csi-driver/provider/v1alpha1"
+)
+
+const (
+	maxFileNameLength = 255
+	maxPathLength     = 4096
+)
+
+// Validate ensures the payload file paths are well formatted.
+func Validate(payloads []*v1alpha1.File) error {
+	for i := range payloads {
+		if err := validatePath(payloads[i].Path); err != nil {
+			return err
+		}
+		if filepath.Clean(payloads[i].Path) != payloads[i].Path {
+			return fmt.Errorf("invalid filepath: %q", payloads[i].Path)
+		}
+	}
+
+	return nil
+}
+
+// WritePayloads writes the files to target directory. One of the main
+// deviations from atomic_writer.go is that this does not attempt to be atomic
+// in updates. atomic_writer.go relies upon ../data folder, but in CSI driver
+// context that path is outside of the tmpfs.
+func WritePayloads(path string, payloads []*v1alpha1.File) error {
+	for _, payload := range payloads {
+		content := payload.Contents
+		mode := os.FileMode(payload.Mode)
+		fullPath := filepath.Join(path, payload.Path)
+		baseDir, _ := filepath.Split(fullPath)
+
+		if err := os.MkdirAll(baseDir, os.ModePerm); err != nil {
+			return fmt.Errorf("unable to create directory %s: %w", baseDir, err)
+		}
+
+		if err := os.WriteFile(fullPath, content, mode); err != nil {
+			return fmt.Errorf("unable to write file %s with mode %v: %w", fullPath, mode, err)
+		}
+
+		// Chmod is needed because ioutil.WriteFile() ends up calling
+		// open(2) to create the file, so the final mode used is "mode &
+		// ~umask". But we want to make sure the specified mode is used
+		// in the file no matter what the umask is.
+		if err := os.Chmod(fullPath, mode); err != nil {
+			return fmt.Errorf("unable to change file %s with mode %v: %w", fullPath, mode, err)
+		}
+	}
+	return nil
+}
+
+// validatePath validates a single path, returning an error if the path is
+// invalid.  paths may not:
+//
+// 1. be absolute
+// 2. contain '..' as an element
+// 3. start with '..'
+// 4. contain filenames larger than 255 characters
+// 5. be longer than 4096 characters
+func validatePath(targetPath string) error {
+	// TODO: somehow unify this with the similar api validation,
+	// validateVolumeSourcePath; the error semantics are just different enough
+	// from this that it was time-prohibitive trying to find the right
+	// refactoring to re-use.
+	if targetPath == "" {
+		return fmt.Errorf("invalid path: must not be empty: %q", targetPath)
+	}
+	if path.IsAbs(targetPath) {
```

Comment:
We should use `filepath` so this is valid for windows too

Reply:
this was from https://github.com/kubernetes/kubernetes/blob/9c4ffa7e11bb2d0beb25ef1cac6e0699acfc350a/pkg/volume/util/atomic_writer.go#L256

Looking at path.IsAbs and filepath.IsAbs I'm fairly certain that their implementations are equivalent, so I'll move to `filepath` as that seems more semantically correct. Think we should propose a change to automic_writer.go too?

Reply:
Yeah, we should propose a change to `atomic_writer.go`. We can open a PR and get feedback from the community.